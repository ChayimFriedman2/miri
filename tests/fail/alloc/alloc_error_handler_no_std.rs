//@compile-flags: -Cpanic=abort
#![feature(start, core_intrinsics)]
#![feature(alloc_error_handler)]
#![feature(allocator_api)]
#![no_std]

extern crate alloc;

use alloc::alloc::*;
use alloc::boxed::Box;
use core::ptr::NonNull;

struct BadAlloc;

// Create a failing allocator; that is the only way to actually call the alloc error handler.
unsafe impl Allocator for BadAlloc {
    fn allocate(&self, _l: Layout) -> Result<NonNull<[u8]>, AllocError> {
        Err(AllocError)
    }

    unsafe fn deallocate(&self, _ptr: NonNull<u8>, _layout: Layout) {
        unreachable!();
    }
}

#[alloc_error_handler]
fn alloc_error_handler(_: Layout) -> ! {
    extern "Rust" {
        fn miri_write_to_stderr(bytes: &[u8]);
    }
    let msg = "custom alloc error handler called!\n";
    unsafe { miri_write_to_stderr(msg.as_bytes()) };
    core::intrinsics::abort(); //~ERROR: aborted
}

// rustc requires us to provide some more things that aren't actually used by this test
mod plumbing {
    use super::*;

    #[panic_handler]
    fn panic_handler(_: &core::panic::PanicInfo) -> ! {
        core::intrinsics::abort();
    }

    struct NoAlloc;

    unsafe impl GlobalAlloc for NoAlloc {
        unsafe fn alloc(&self, _: Layout) -> *mut u8 {
            unreachable!();
        }

        unsafe fn dealloc(&self, _: *mut u8, _: Layout) {
            unreachable!();
        }
    }

    #[global_allocator]
    static GLOBAL: NoAlloc = NoAlloc;
}

#[start]
fn start(_: isize, _: *const *const u8) -> isize {
    let _b = Box::new_in(0, BadAlloc);
    0
}
