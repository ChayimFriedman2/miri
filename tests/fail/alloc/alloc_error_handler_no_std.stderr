custom alloc error handler called!
error: abnormal termination: the program aborted execution
  --> $DIR/alloc_error_handler_no_std.rs:LL:CC
   |
LL |     core::intrinsics::abort();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^ the program aborted execution
   |
   = note: BACKTRACE:
   = note: inside `alloc_error_handler` at $DIR/alloc_error_handler_no_std.rs:LL:CC
note: inside `_::__rg_oom`
  --> $DIR/alloc_error_handler_no_std.rs:LL:CC
   |
LL | #[alloc_error_handler]
   | ---------------------- in this procedural macro expansion
LL | fn alloc_error_handler(_: Layout) -> ! {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = note: inside `alloc::alloc::handle_alloc_error::rt_error` at RUSTLIB/alloc/src/alloc.rs:LL:CC
   = note: inside `alloc::alloc::handle_alloc_error` at RUSTLIB/alloc/src/alloc.rs:LL:CC
   = note: inside `alloc::boxed::Box::<i32, BadAlloc>::new_uninit_in` at RUSTLIB/alloc/src/boxed.rs:LL:CC
   = note: inside `alloc::boxed::Box::<i32, BadAlloc>::new_in` at RUSTLIB/alloc/src/boxed.rs:LL:CC
note: inside `start`
  --> $DIR/alloc_error_handler_no_std.rs:LL:CC
   |
LL |     let _b = Box::new_in(0, BadAlloc);
   |              ^^^^^^^^^^^^^^^^^^^^^^^^
   = note: this error originates in the attribute macro `alloc_error_handler` (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to 1 previous error

